### ISSUE TITLE
Can't build on Windows with MinGW

### ISSUE THREAD

**Issue by kewp** [mfem/mfem#199]
I've used the latest MinGW-W64 install, set the PATH, and ran

`make serial -j`

and got a whole bunch of 'notes' (can't find where the actual error is ... ?)

> mesh/ncmesh.cpp:3264:10: note: in expansion of macro 'MFEM_VERIFY'
          MFEM_VERIFY(child, "element " << id << " cannot have two parents.");
          ^~~~~~~~~~~
mesh/../general/error.hpp:38:25: note: suggested alternative: '__FLT_DIG__'
 #define _MFEM_FUNC_NAME __FUNCSIG__
                         ^
mesh/../general/error.hpp:43:30: note: in expansion of macro '_MFEM_FUNC_NAME'
    "\n ... in function: " << _MFEM_FUNC_NAME << \
                              ^~~~~~~~~~~~~~~
mesh/../general/error.hpp:52:31: note: in expansion of macro 'MFEM_LOCATION'
       mfemMsgStream << msg << MFEM_LOCATION;                            \
                               ^~~~~~~~~~~~~
mesh/../general/error.hpp:70:7: note: in expansion of macro '_MFEM_MESSAGE'
       _MFEM_MESSAGE("Verification failed: ("            \
       ^~~~~~~~~~~~~
mesh/ncmesh.cpp:3264:10: note: in expansion of macro 'MFEM_VERIFY'
          MFEM_VERIFY(child, "element " << id << " cannot have two parents.");
          ^~~~~~~~~~~
make[1]: *** [fem/datacollection.o] Error 1
make[1]: *** [mesh/ncmesh.o] Error 1
make[1]: *** wait: No children.  Stop.
make: *** [serial] Error 2

**Comment by v-dobrev** [comment]
Hi @kewp,

Thanks for reporting this issue. The easiest way to resolve it is to overwrite
```c++
#define _MFEM_FUNC_NAME __FUNCSIG__
```
with
```c++
#define _MFEM_FUNC_NAME __PRETTY_FUNCTION__
```
or
```c++
#define _MFEM_FUNC_NAME __func__
```
We can fix the issue in MFEM, however I don't know what macros are defined by MinGW-W64 that will distinguish it from, say, the Visual Studio compiler. Also, I'm not sure which function-name macro is supported by MinGW-W64: is it `__PRETTY_FUNCTION__`, `__func__`, or something else.

Let us know what you find and we can create a pull request with a fix, or if you'd like - you can create a pull request.

Thanks,
Veselin

**Comment by kewp** [comment]
The errors stopped coming but then, after a long pause, it exited with an error but no message.

> g++  -O3   -c fem/pnonlinearform.cpp -o fem/pnonlinearform.o
g++  -O3   -c fem/sidredatacollection.cpp -o fem/sidredatacollection.o
g++  -O3   -c fem/staticcond.cpp -o fem/staticcond.o
make[1]: *** wait: No children.  Stop.
make[1]: *** Waiting for unfinished jobs....
make[1]: *** wait: No children.  Stop.
make: *** [serial] Error 2

c:\Users\karl\tst\mfem>

**Comment by tzanio** [comment]
@kewp, try removing the `-j`, i.e. just `make serial`.

**Comment by kewp** [comment]
Not sure what I'm doing differently. On a new machine I did a clean download and build and get a `CreateProcess` error.

> C:\Users\Karl\Downloads\mfem-3.3>mingw32-make serial
process_begin: CreateProcess(NULL, uname -s, ...) failed.
mingw32-make: config/defaults.mk:19: pipe: No error
mingw32-make -f makefile config MFEM_USE_MPI=NO MFEM_DEBUG=NO
mingw32-make[1]: Entering directory 'C:/Users/Karl/Downloads/mfem-3.3'
process_begin: CreateProcess(NULL, uname -s, ...) failed.
mingw32-make[1]: config/defaults.mk:19: pipe: No error
mingw32-make -C config all
process_begin: CreateProcess(NULL, uname -snm, ...) failed.
mingw32-make[1]: config/defaults.mk:199: pipe: No such file or directory
mingw32-make[2]: Entering directory 'C:/Users/Karl/Downloads/mfem-3.3/config'
def was unexpected at this time.
makefile:39: recipe for target 'header' failed
mingw32-make[2]: *** [header] Error 255
mingw32-make[2]: Leaving directory 'C:/Users/Karl/Downloads/mfem-3.3/config'
makefile:368: recipe for target 'local-config' failed
mingw32-make[1]: *** [local-config] Error 2
mingw32-make[1]: Leaving directory 'C:/Users/Karl/Downloads/mfem-3.3'
makefile:299: recipe for target 'serial' failed
mingw32-make: *** [serial] Error 2

C:\Users\Karl\Downloads\mfem-3.3>

**Comment by v-dobrev** [comment]
I do not know what the issue is, I'm not familiar with `mingw32`. You can try the CMake build system, see [https://cmake.org/](https://cmake.org/) - it seems that it supports output for `mingw32-make`: [https://cmake.org/cmake/help/v3.8/manual/cmake-generators.7.html#makefile-generators](https://cmake.org/cmake/help/v3.8/manual/cmake-generators.7.html#makefile-generators)

**Comment by kewp** [comment]
Couldn't get cmake to work either

> C:\Users\Karl\Downloads\mfem-3.3>cmake -G "MinGW Makefiles"
C:\Users\Karl\Downloads\mfem-3.3>cmake -G "MinGW Makefiles"
-- (optional) USER_CONFIG = C:/Users/Karl/Downloads/mfem-3.3/config/user.cmake
-- The CXX compiler identification is GNU 7.1.0
-- Check for working CXX compiler: C:/mingw-w64/i686-7.1.0-posix-dwarf-rt_v5-rev0/mingw32/bin/g++.exe
-- Check for working CXX compiler: C:/mingw-w64/i686-7.1.0-posix-dwarf-rt_v5-rev0/mingw32/bin/g++.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:46 (message):
  MFEM does not support in-source CMake builds at this time.


-- Configuring incomplete, errors occurred!
See also "C:/Users/Karl/Downloads/mfem-3.3/CMakeFiles/CMakeOutput.log".

**Comment by trb413** [comment]
You cannot run CMake in the source directory. Instead, create a new directory, 'build' or something, and call CMake from that directory with the syntax

cmake /path/to/mfem/source -G "MinGW Makefiles"

Based on your posted output, I believe that should be

cmake C:\Users\Karl\Downloads\mfem-3.3 -G "MinGW Makefiles"

Good luck!

**Comment by kewp** [comment]
That worked :)

> [ 78%] Building CXX object CMakeFiles/mfem.dir/fem/fespace.cpp.obj
[ 80%] Building CXX object CMakeFiles/mfem.dir/fem/geom.cpp.obj
[ 81%] Building CXX object CMakeFiles/mfem.dir/fem/gridfunc.cpp.obj
[ 83%] Building CXX object CMakeFiles/mfem.dir/fem/hybridization.cpp.obj
[ 85%] Building CXX object CMakeFiles/mfem.dir/fem/intrules.cpp.obj
[ 87%] Building CXX object CMakeFiles/mfem.dir/fem/linearform.cpp.obj
[ 89%] Building CXX object CMakeFiles/mfem.dir/fem/lininteg.cpp.obj
[ 90%] Building CXX object CMakeFiles/mfem.dir/fem/nonlinearform.cpp.obj
[ 92%] Building CXX object CMakeFiles/mfem.dir/fem/nonlininteg.cpp.obj
[ 94%] Building CXX object CMakeFiles/mfem.dir/fem/staticcond.cpp.obj
[ 96%] Linking CXX static library libmfem.a
[100%] Built target mfem

C:\Users\Karl\Downloads\mfem-3.3\build>

**Comment by kewp** [comment]
Problems now when building the examples

> Scanning dependencies of target exec_prerequisites
[  0%] Built target exec_prerequisites
[ 63%] Built target mfem
Scanning dependencies of target ex9
[ 64%] Building CXX object examples/CMakeFiles/ex9.dir/ex9.cpp.obj
[ 65%] Linking CXX executable ex9.exe
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x17): undefined reference to `_imp__send@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0xb6): undefined reference to `_imp__recv@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x178): undefined reference to `_imp__recv@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x207): undefined reference to `_imp__send@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x2dc): undefined reference to `_imp__send@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x392): undefined reference to `_imp__send@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x3e5): undefined reference to `_imp__closesocket@4'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x4bb): undefined reference to `_imp__send@16'
../libmfem.a(socketstream.cpp.obj):socketstream.cpp:(.text+0x4dd): undefined reference to `

**Comment by v-dobrev** [comment]
It looks like the progma
```c++
#pragma comment(lib, "ws2_32.lib")
```
from [general/socketstream.cpp:33](https://github.com/mfem/mfem/blob/master/general/socketstream.cpp#L33) is not sufficient to enable linking with `ws2_32.lib` when using MinGW. Try adding the following lines after line [CMakeLists.txt:272](https://github.com/mfem/mfem/blob/master/CMakeLists.txt#L272):
```cmake
if (MINGW)
   target_link_libraries(mfem ws2_32)
endif()
```
If that does not work try changing `ws2_32` with  the full path to the library, including the extension as well.

**Comment by kewp** [comment]
Almost there. Examples build to 96%

> Scanning dependencies of target ex16
[ 91%] Building CXX object examples/CMakeFiles/ex16.dir/ex16.cpp.obj
[ 93%] Linking CXX executable ex16.exe
[ 93%] Built target ex16
Scanning dependencies of target ex17
[ 94%] Building CXX object examples/CMakeFiles/ex17.dir/ex17.cpp.obj
[ 95%] Linking CXX executable ex17.exe
[ 95%] Built target ex17
Scanning dependencies of target sundials_ex9
[ 96%] Building CXX object examples/sundials/CMakeFiles/sundials_ex9.dir/ex9.cpp.obj
C:\Users\karl\Downloads\mfem-3.3\examples\sundials\ex9.cpp:34:2: error: #error This example requires that MFEM is built with MFEM_USE_SUNDIALS=YES
 #error This example requires that MFEM is built with MFEM_USE_SUNDIALS=YES
  ^~~~~
C:\Users\karl\Downloads\mfem-3.3\examples\sundials\ex9.cpp: In function 'int main(int, char**)':
C:\Users\karl\Downloads\mfem-3.3\examples\sundials\ex9.cpp:148:4: error: 'CVODESolver' was not declared in this scope
    CVODESolver *cvode = NULL;
    ^~~~~~~~~~~

**Comment by kewp** [comment]
Examples don't run though ...

http://i.imgur.com/9OC4voP.png

**Comment by v-dobrev** [comment]
I created a test branch for building under windows with MinGW: [mingw-build-fix](https://github.com/mfem/mfem/tree/mingw-build-fix). I just tried it with the latest MinGW and it seems to work fine. Here is what I did after downloading the branch:
```sh
cd mfem
mkdir build
cd build
cmake .. -G "MinGW Makefiles"
cmake --build .
cmake --build . --target exec
cmake --build . --target test
```
The `exec` target builds all miniapps and examples, and the `test` target runs them. The only test that failed for me is `performance_ex1` (i.e. `miniapps/performance/ex1.cpp`) - there is some issue when running that example under windows, it happens when building with visual studio as well.
